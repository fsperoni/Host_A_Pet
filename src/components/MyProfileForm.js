import React, { useState, useContext } from "react";
import { useNavigate } from "react-router-dom";
import Alert from "./Alert";
import HostAPetApi from "./Api";
import UserContext from "../hooks/useUserContext";

/** User Profile edit form. */
const MyProfileForm = ({ logout }) => {
  const history = useNavigate();
  const { currentUser, setCurrentUser } = useContext(UserContext);
  const [formData, setFormData] = useState({
    username: currentUser.username, 
    firstName: currentUser.firstName,
    lastName: currentUser.lastName, 
    email: currentUser.email,
    phone: currentUser.phone,
    postalCode: currentUser.postalCode,
    password: "",
  });
  const [formErrors, setFormErrors] = useState([]);
  const [isSaved, setIsSaved] = useState(false);

  const handleSubmit = async (evt) => {
    evt.preventDefault();
    const userData = {
      firstName: formData.firstName,
      lastName: formData.lastName,
      email: formData.email,
      phone: formData.phone,
      postalCode: formData.postalCode,
      password: formData.password
    };

    const username = formData.username;
    let updatedUser;

    try {
      updatedUser = await HostAPetApi.updateUser(username, userData);
    } catch (errors) {
      setFormErrors(errors);
      return;
    }

    setFormData(f => ({ ...f, password: "" }));
    setFormErrors([]);
    setIsSaved(true);

    setCurrentUser(updatedUser);
  }

  /** Handle form field changes */
  const handleChange = (evt) => {
    const { name, value } = evt.target;
    setFormData(form => ({ ...form, [name]: value }));
    setFormErrors([]);
  }
  /** Handle form cancel button */
  const handleCancel = (evt) => {
    evt.preventDefault();
    history("/");
  }

  /** Handle form delete button */
  const handleDelete = async (evt) => {
    evt.preventDefault();
    const username = formData.username;
    let res;
    try {
      res = await HostAPetApi.deleteUser(username);
    } catch (errors) {
      setFormErrors(errors);
      return;
    }

    setFormData(f => ({ ...f, password: "" }));
    setFormErrors([]);
    history("/");
    logout();
  }

  return (
    <div className="col-md-6 col-lg-4 offset-md-3 offset-lg-4">
      <h3>Profile</h3>
      <div className="card">
        <div className="card-body">
          <form>
            <div className="form-group">
              <label>Username:</label>
              <p className="form-control-plaintext">{formData.username}</p>
            </div>
            <div className="form-group">
              <label>First Name:</label>
              <input
                name="firstName"
                className="form-control"
                value={formData.firstName}
                onChange={handleChange}
              />
            </div>
            <div className="form-group">
              <label>Last Name:</label>
              <input
                name="lastName"
                className="form-control"
                value={formData.lastName}
                onChange={handleChange}
              />
            </div>
            <div className="form-group">
              <label>Email:</label>
              <input
                name="email"
                className="form-control"
                value={formData.email}
                onChange={handleChange}
              />
            </div>
            <div className="form-group">
              <label>Phone:</label>
              <input
                name="phone"
                className="form-control"
                value={formData.phone}
                onChange={handleChange}
              />
            </div>
            <div className="form-group">
              <label>Postal code:</label>
              <input
                name="postalCode"
                className="form-control"
                value={formData.postalCode}
                onChange={handleChange}
              />
            </div>
            <div className="form-group">
              <label>Confirm password to make changes:</label>
              <input
                type="password"
                name="password"
                className="form-control"
                value={formData.password}
                onChange={handleChange}
              />
            </div>

            {formErrors.length
              ? <Alert type="danger" messages={formErrors} />
              : null}

            {isSaved
              ?
              <Alert type="success" messages={["Updated successfully."]} />
              : null}
            <div className="container mt-4">
              <button
                className="btn btn-primary btn-block"
                onClick={handleSubmit}
              >
                Save Changes
              </button>
              <button
                className="btn btn-warning btn-block ms-4"
                onClick={handleCancel}
              >
                Cancel
              </button>
              <button
                className="btn btn-danger btn-block ms-4"
                onClick={handleDelete}
              >
                Delete Account
              </button>
            </div>
          </form>
        </div>
      </div>
    </div>
  );
}

export default MyProfileForm;
