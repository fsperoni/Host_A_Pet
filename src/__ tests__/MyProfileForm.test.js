import { render } from "@testing-library/react";
import { BrowserRouter } from "react-router-dom";
import MyProfileForm from "../components/MyProfileForm";
import UserContext from "../hooks/useUserContext";

const {
  currentAdminUser,
} = require("./_testCommon");

// smoke test
it('renders without crashing', function () {
  render(
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <MyProfileForm logout={function () { }} />
      </UserContext.Provider>
    </BrowserRouter>
  );
});

//snapshot test
it("matches snapshot", function () {
  const { asFragment } = render(
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <MyProfileForm logout={function () { }} />
      </UserContext.Provider>
    </BrowserRouter>
  );
  expect(asFragment()).toMatchSnapshot();
});

it("should have 'Profile' in the component.", function () {
  const { getByText } = render(
    <BrowserRouter>
      <UserContext.Provider
        value={{ currentUser: currentAdminUser(), function() { } }}>
        <MyProfileForm logout={function () { }} />
      </UserContext.Provider>
    </BrowserRouter>
  );
  expect(getByText("Profile")).toBeInTheDocument();
});
