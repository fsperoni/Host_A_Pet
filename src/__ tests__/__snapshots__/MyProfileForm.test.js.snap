// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="col-md-6 col-lg-4 offset-md-3 offset-lg-4 mb-3"
  >
    <h3>
      Profile
    </h3>
    <div
      class="card"
    >
      <div
        class="card-body"
      >
        <form>
          <div
            class="form-group"
          >
            <label>
              First Name:
            </label>
            <input
              class="form-control"
              name="firstName"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label>
              Last Name:
            </label>
            <input
              class="form-control"
              name="lastName"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label>
              Email:
            </label>
            <input
              class="form-control"
              name="email"
              value="noemail@example.com"
            />
          </div>
          <div
            class="form-group"
          >
            <label>
              Phone:
            </label>
            <input
              class="form-control"
              name="phone"
              value="555-123-4567"
            />
          </div>
          <div
            class="form-group"
          >
            <label>
              Postal code:
            </label>
            <input
              class="form-control"
              name="postalCode"
              value=""
            />
          </div>
          <div
            class="form-group"
          >
            <label>
              Confirm password to make changes:
            </label>
            <input
              class="form-control"
              name="password"
              type="password"
              value=""
            />
          </div>
          <div
            class="container mt-4"
          >
            <button
              class="btn btn-primary btn-block"
            >
              Save Changes
            </button>
            <button
              class="btn btn-warning btn-block ms-4"
            >
              Cancel
            </button>
            <button
              class="btn btn-danger btn-block ms-4"
            >
              Delete Account
            </button>
          </div>
        </form>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
